opm/autodiff/NewtonSolver.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+  Copyright 2015 SINTEF ICT, Applied Mathematics.
+  Copyright 2015 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_NEWTONSOLVER_HEADER_INCLUDED
+#define OPM_NEWTONSOLVER_HEADER_INCLUDED
+
+#include <opm/autodiff/AutoDiffBlock.hpp>
+#include <opm/core/utility/parameters/ParameterGroup.hpp>
+
+namespace Opm {
+
+
+    /// A Newton solver class suitable for general fully-implicit models.
+    template <class PhysicalModel>
+    class NewtonSolver
+    {
+    public:
+        // ---------  Types and enums  ---------
+        typedef AutoDiffBlock<double> ADB;
+        typedef ADB::V V;
+        typedef ADB::M M;
+
+        // The Newton relaxation scheme type
+        enum RelaxType { DAMPEN, SOR };
+
+        // Solver parameters controlling nonlinear Newton process.
+        struct SolverParameter
+        {
+            enum RelaxType relax_type_;
+            double         relax_max_;
+            double         relax_increment_;
+            double         relax_rel_tol_;
+            int            max_iter_; // max newton iterations
+            int            min_iter_; // min newton iterations
+
+            SolverParameter( const parameter::ParameterGroup& param );
+            SolverParameter();
+
+            void reset();
+        };
+
+        // Forwarding types from PhysicalModel.
+        typedef typename PhysicalModel::ReservoirState ReservoirState;
+        typedef typename PhysicalModel::WellState WellState;
```

Comment:
Yes, I agree (if you mean having a single state class instead of one for the reservoir and one for the wells). However, I want to do this process one step at a time, so I'd not change that in the current PR.